src/search.cpp

```diff
@@ -209,17 +209,25 @@ void GreedySearch_Cpu::SampleTopKTopP(int k, float p, float temperature) {
       continue;
     }
     std::span<float> const scores = next_token_scores_.CpuSpan().subspan(batch_id * params_->config.model.vocab_size, params_->config.model.vocab_size);
-    SoftMax(scores, temperature);
     // Find the top K scores
     std::vector<int> indices(scores.size());
     std::iota(indices.begin(), indices.end(), 0);
     std::partial_sort(indices.begin(), indices.begin() + k, indices.end(), [scores = scores.data()](int i, int j) { return scores[i] > scores[j]; });
+    std::vector<float> scores_top_k(k, 0.0f);
+    for (int i = 0; i < k; i++) {
+      scores_top_k[i] = scores[indices[i]];
+    }
+    if (k > 1) {
+      SoftMaxWithMax(scores_top_k, temperature, scores_top_k[0]);
+    } else {
+      scores_top_k[0] = 1.0f;
+    }
     // Sample a probability threshold
     float threshold = dis(gen_);
-    int32_t token = indices[k - 1];
+    int32_t token = indices[0];
```

Comment:
Suggest to use k - 1 instead of 0 here. Also change loop to `for (int i = 0; i < k - 1; i++)`

Reply:
done.